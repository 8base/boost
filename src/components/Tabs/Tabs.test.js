// @flow
import React from 'react';
import { Tabs } from './';

describe('<Tabs />', () => {

  it('should render Tabs', () => {
    const wrapper = mount(
      <Tabs defaultSelectedTabId="1">
        <Tabs.Title tabId="1"><div>Tab-1</div></Tabs.Title>
        <Tabs.Title tabId="2"><div>Tab-2</div></Tabs.Title>
        <Tabs.Panel tabId="1">
          <div className="body-1">Body-1</div>
        </Tabs.Panel>
        <Tabs.Panel tabId="2">
          <div className="body-2">Body-2</div>
        </Tabs.Panel>
      </Tabs>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render selected tab', () => {
    const wrapper = mount(
      <Tabs selectedTabId="1">
        <Tabs.Title tabId="1"><div>Tab-1</div></Tabs.Title>
        <Tabs.Title tabId="2"><div>Tab-2</div></Tabs.Title>
        <Tabs.Panel tabId="1">
          <div className="body-1">Body-1</div>
        </Tabs.Panel>
        <Tabs.Panel tabId="2">
          <div className="body-2">Body-2</div>
        </Tabs.Panel>
      </Tabs>,
    );

    expect(wrapper.find('.body-1').text()).toBe('Body-1');
    expect(wrapper.find('.body-2')).toHaveLength(0);

    wrapper.setProps({ selectedTabId: '2' });

    expect(wrapper.find('.body-1')).toHaveLength(0);
    expect(wrapper.find('.body-2').text()).toBe('Body-2');
  });


  it('should render head as a render prop', () => {
    const wrapper = mount(
      <Tabs selectedTabId="1">
        <Tabs.Title tabId="1">{ ({ selected }) => <div className="head-1">Tab-1 is { selected && 'selected' }</div> }</Tabs.Title>
        <Tabs.Title tabId="2">{ ({ selected }) => <div className="head-2">Tab-2 is { !selected && 'not selected' }</div> }</Tabs.Title>
        <Tabs.Panel tabId="1">Body-1</Tabs.Panel>
        <Tabs.Panel tabId="2">Body-2</Tabs.Panel>
      </Tabs>,
    );

    expect(wrapper.find('.head-1').text()).toBe('Tab-1 is selected');
    expect(wrapper.find('.head-2').text()).toBe('Tab-2 is not selected');
  });


  it('should call onSelect callback', () => {
    const onSelect = jest.fn();
    const wrapper = mount(
      <Tabs selectedTabId="1" onSelect={ onSelect }>
        <Tabs.Title tabId="1"><div>Tab-1</div></Tabs.Title>
        <Tabs.Title tabId="2"><div>Tab-2</div></Tabs.Title>
        <Tabs.Panel tabId="1">
          <div className="body-1">Body-1</div>
        </Tabs.Panel>
        <Tabs.Panel tabId="2">
          <div className="body-2">Body-2</div>
        </Tabs.Panel>
      </Tabs>,
    );

    wrapper.find(Tabs.Title).at(0).simulate('click');
    expect(onSelect.mock.calls[0][0]).toBe('1');

    wrapper.find(Tabs.Title).at(1).simulate('click');
    expect(onSelect.mock.calls[1][0]).toBe('2');
  });


  it('should toggle tabs in statefull mode', () => {
    const wrapper = mount(
      <Tabs defaultSelectedTabId="1">
        <Tabs.Title tabId="1"><div>Tab-1</div></Tabs.Title>
        <Tabs.Title tabId="2"><div>Tab-2</div></Tabs.Title>
        <Tabs.Panel tabId="1">
          <div className="body-1">Body-1</div>
        </Tabs.Panel>
        <Tabs.Panel tabId="2">
          <div className="body-2">Body-2</div>
        </Tabs.Panel>
      </Tabs>,
    );

    wrapper.find(Tabs.Title).at(1).simulate('click');
    expect(wrapper.find('.body-1')).toHaveLength(0);
    expect(wrapper.find('.body-2').text()).toBe('Body-2');

    wrapper.find(Tabs.Title).at(0).simulate('click');
    expect(wrapper.find('.body-1').text()).toBe('Body-1');
    expect(wrapper.find('.body-2')).toHaveLength(0);
  });


  it('should render panels with force', () => {
    const onSelect = jest.fn();
    const wrapper = mount(
      <Tabs selectedTabId="1" onSelect={ onSelect }>
        <Tabs.Title tabId="1"><div>Tab-1</div></Tabs.Title>
        <Tabs.Title tabId="2"><div>Tab-2</div></Tabs.Title>
        <Tabs.Panel tabId="1" forceRender>
          <span className="body-1">Body-1</span>
        </Tabs.Panel>
        <Tabs.Panel tabId="2" forceRender>
          <span className="body-2">Body-2</span>
        </Tabs.Panel>
      </Tabs>,
    );

    expect(wrapper.find('.body-1').text()).toBe('Body-1');
    expect(wrapper.find('.body-2').text()).toBe('Body-2');
  });
});

