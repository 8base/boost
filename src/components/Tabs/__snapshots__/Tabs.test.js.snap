// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should render Tabs 1`] = `
<Tabs
  defaultSelectedTabId="1"
>
  <branch(Component)
    defaultSelectedTabId="1"
  >
    <withStateHandlers(Component)
      defaultSelectedTabId="1"
    >
      <Component
        defaultSelectedTabId="1"
        onSelect={[Function]}
        selectedTabId="1"
      >
        <TabTitle
          tabId="1"
        >
          <TabTitleBase
            tabId="1"
            tabs={
              Object {
                "onSelect": [Function],
                "selectedTabId": "1",
              }
            }
          >
            <Boost(tabTitle)
              onClick={[Function]}
              tagName="div"
            >
              <Tag
                className="css-rkvmzw-StyledTag emotion-0"
                onClick={[Function]}
                tagName="div"
              >
                <div
                  className="css-rkvmzw-StyledTag emotion-0"
                  onClick={[Function]}
                >
                  <div>
                    Tab-1
                  </div>
                </div>
              </Tag>
            </Boost(tabTitle)>
          </TabTitleBase>
        </TabTitle>
        <TabTitle
          tabId="2"
        >
          <TabTitleBase
            tabId="2"
            tabs={
              Object {
                "onSelect": [Function],
                "selectedTabId": "1",
              }
            }
          >
            <Boost(tabTitle)
              onClick={[Function]}
              tagName="div"
            >
              <Tag
                className="css-rkvmzw-StyledTag emotion-0"
                onClick={[Function]}
                tagName="div"
              >
                <div
                  className="css-rkvmzw-StyledTag emotion-0"
                  onClick={[Function]}
                >
                  <div>
                    Tab-2
                  </div>
                </div>
              </Tag>
            </Boost(tabTitle)>
          </TabTitleBase>
        </TabTitle>
        <TabPanel
          tabId="1"
        >
          <TabPanelBase
            forceRender={false}
            selected={false}
            tabId="1"
            tabs={
              Object {
                "onSelect": [Function],
                "selectedTabId": "1",
              }
            }
          >
            <Boost(tabPlate)
              selected={true}
              tagName="div"
            >
              <Tag
                className="css-rkvmzw-StyledTag emotion-0"
                selected={true}
                tagName="div"
              >
                <div
                  className="css-rkvmzw-StyledTag emotion-0"
                >
                  <div
                    className="body-1"
                  >
                    Body-1
                  </div>
                </div>
              </Tag>
            </Boost(tabPlate)>
          </TabPanelBase>
        </TabPanel>
        <TabPanel
          tabId="2"
        >
          <TabPanelBase
            forceRender={false}
            selected={false}
            tabId="2"
            tabs={
              Object {
                "onSelect": [Function],
                "selectedTabId": "1",
              }
            }
          />
        </TabPanel>
      </Component>
    </withStateHandlers(Component)>
  </branch(Component)>
</Tabs>
`;
